src/services/FileSystemService.ts

```diff
@@ -1,104 +1,122 @@
 import fs from 'fs';
 import * as nodePath from 'path';
-import { ErrorFactory } from '../lib/errors/ErrorFactory';
 import { injectable } from 'inversify';
 import { IProjectFilesBrowserService, Metadata, MetadataType } from './model';
-
+import { IFs, createFsFromVolume } from 'memfs';
+import { Volume as VSVolume, DirectoryJSON } from 'memfs/lib/volume';
+import { ErrorFactory } from '../lib/errors';
+
+/**
+ * Service for file system browsing
+ *  - uses fs by default
+ *  - can work just in memory with memfs
+ */
 @injectable()
 export class FileSystemService implements IProjectFilesBrowserService {
+  protected fileSystem: IFs | (typeof fs);
+  private virtualVolume: VSVolume | undefined;
+
+  constructor({ isVirtual = false } = {}) {
```

Comment:
One last thing... I admit that `isVirtual` was my idea... just another idea came to my mind... O:-)

I think that in FileSystemService tests, we should rather use `fs.readFile` to check whether e.g. a file has been created rather than using `FileSystemService.readFile`. As well, in the tests of an virtual instance, we should rather use `vfs.readFile`. That will make us sure, that a "real" FSS really writes to the file system as well as that a "virtual" FSS writes to the virtual file system (and not to the real one).

However, I do not insist that you should do that now because your PR did not touch this tests logic. But I think that it would be nice, if we change the constructor to `constructor({ volume?: VSVolume })` (or just `constructor(volume?: VSVolume)`) **now** so that we can change the tests of virtual FSS to test against the volume **later**. What do you think?

Reply:
It's quite fine idea! :)